Added the functionality of Preserved Comments in cupsd.conf when cups…

Ankit3002 opened the pull request:
…ctl is used with command line arguments

Ankit3002 commented:
@zdohnal I made the changes as you requested. Now I 'm printing the comment whenever I read it from cupsd.conf file. I delete the most recent commits and erroneously drop the last PR. Please review this PR and do let me know if these changes are fine or not ?

zdohnal commented on cups/adminutil.c:
```diff
@@ -698,8 +700,11 @@ cupsAdminSetServerSettings(
     server_port = IPP_PORT;
 
-  while (cupsFileGetConf(cupsd, line, sizeof(line), &value, &linenum))
+  while (_cupsFileGetConfAndComments(cupsd, line, sizeof(line), &value, &linenum, comment, sizeof(comment)))
```
What about returning the comment via ```line``` parameter and remove the new additional parameters? Then the comment printout can happen via last cupsFilePrintf(). WDYT?

zdohnal commented on cups/file.c:
```diff
@@ -841,6 +841,142 @@ cupsFileGetConf(cups_file_t *fp,	/* I  - CUPS file */
+    */
+
+    if ((ptr = strchr(buf, '#')) != NULL)
+    {
```
If we go with passing comments via ```line```, we can detect the comment, then remove the leading and trailing whitespaces and return ```buf```.

Ankit3002 replied:
@zdohnal I made the changes as you requested. Now I 'm using line variable to get comments. Please review them.

zdohnal reviewed (changes requested):
IMO we can simplify the PR a little bit more - we can get comments via ```line``` itself.

zdohnal commented on cups/file.c:
```diff
@@ -769,6 +769,7 @@ cupsFileGetConf(cups_file_t *fp,	/* I  - CUPS file */
+    return (buf);
 	*ptr = '\0';
       }
     }
```
Additionally you should review every cupsFilePut() below in ```cupsAdminSetServerSettings()``` which add a comment - once we start supporting comments, we can't put new comments in.

Rephrase the comments mentioned in ```cupsFilePut()``` and put it to the proper place at ```conf/cupsd.conf.in```, f.e. when line contains "</Location>" and we are in /admin location:

The code:
```
  if (remote_admin)
          cupsFilePuts(temp, "  # Allow remote administration...\n");
  else
          cupsFilePuts(temp, "  # Restrict access to the admin pages...\n");
```

The default cupsd.conf.in:

```
# Restrict access to the admin pages...
<Location /admin>
  AuthType Default
  Require user @SYSTEM
  Order allow,deny
</Location>
```

so once you remove the cupsFilePut(), update the cupsd.conf.in to look like this f.e.:

```
# Access to the admin pages:
# - default action defined by Order', see 'man cupsd.conf'
# - use 'Allow'/'Deny' for configuring access
<Location /admin>
  AuthType Default
  Require user @SYSTEM
  Order allow,deny
</Location>
```

Ankit3002 replied:
`     else if (in_root_location &&
               (remote_admin >= 0 || remote_any >= 0 || share_printers >= 0))
      {
	wrote_root_location = 1;

	if (remote_admin > 0 && share_printers > 0)
          cupsFilePuts(temp, "  # Allow shared printing and remote "
	                     "administration...\n");
	else if (remote_admin > 0)
          cupsFilePuts(temp, "  # Allow remote administration...\n");
	else if (share_printers > 0)
          cupsFilePuts(temp, "  # Allow shared printing...\n");
	else if (remote_any > 0)
          cupsFilePuts(temp, "  # Allow remote access...\n");
	else
          cupsFilePuts(temp, "  # Restrict access to the server...\n");

        cupsFilePuts(temp, "  Order allow,deny\n");

	if (remote_admin > 0 || remote_any > 0 || share_printers > 0)
	{
	  if (remote_any >= 0)
	    cupsFilePrintf(temp, "  Allow %s\n", remote_any > 0 ? "all" : "@LOCAL");
	  else
	    cupsFilePrintf(temp, "  Allow %s\n", old_remote_any > 0 ? "all" : "@LOCAL");
	}
      }`

@zdohnal for the above part what should be the comment that I need to write in conf/cupsd.conf.in ?

zdohnal replied:
For example:

```
# Access to the server root (/):
# - default action defined by 'Order', see 'man cupsd.conf'
# - use 'Allow'/'Deny' for configuring access
# - allowing access is required for printer sharing or remote administration
```

zdohnal reviewed (changes requested):
We can't change the behavior of public function - I'm sorry if my comment sounded like it, I still meant to use a new private function.

Additionally we have to take care of adding comments in ```cupsAdminSetServerSettings()```.

Ankit3002 commented:
@zdohnal I made the changes as you requested. I create a private Api to read comments and lines. 
1. The private api use same line variable to read both comments and configurations.
2. I changed the comments in cupsd.conf.in as well.

Ankit3002 commented on cups/adminutil.c:
```diff
@@ -700,6 +702,16 @@ cupsAdminSetServerSettings(
 
   while (cupsFileGetConf(cupsd, line, sizeof(line), &value, &linenum))
   {
+    if (strchr(line, '#') != NULL)
```
1. linenum checking is used in order to put blank line in between multilines.
2. I write the if statement on the top so that comments got written on top. and the last else block have again check for # so that that else won't write comments. Since line contains both configuration and comments.

Ankit3002 commented:
> @zdohnal I made the changes as you requested. I create a private Api to read comments and lines.
> 
> 1. The private api use same line variable to read both comments and configurations.
> 2. I changed the comments in cupsd.conf.in as well.

@zdohnal Any update on this ?

Ankit3002 commented:
> @zdohnal I made the changes as you requested. I create a private Api to read comments and lines.
> 
>     1. The private api use same line variable to read both comments and configurations.
> 
>     2. I changed the comments in cupsd.conf.in as well.
> 
>     3. I moved the if else block at the end as well. Linenum checking is used only to not include blank lines between multiline comments.

@zdohnal Any update on this ?

zdohnal commented:
Hi @Ankit3002 ,

I'm sorry, I was on vacation - I'll review this today or tomorrow.

Ankit3002 commented:
@zdohnal  cupsd.conf.in file is placed in .gitignore, should I remove it from there ? as I won't be able to push the changes in cupsd.conf.in because of this

Ankit3002 commented:
> Are you sure? I don't see the file in .gitignore.

I 'm sorry , I got confused it with cupsd.conf

zdohnal commented:
> @zdohnal I made the changes as you requested. I create a private Api to read comments and lines.
> 
>     1. The private api is using same line variable to read both comments and configurations.

Ok, please update the comments and debug printf to match the new function - since it was copied from cupsGetConf(), the comments+debugs stayed the same.

> 
>     2. I changed the comments in cupsd.conf.in as well.

cupsAdminSetServerSettings() still puts comments into cupsd.conf file, which we can't allow once we support reading comments in the functions (otherwise we would end up with duplicity and bigger and bigger cupsd.conf with every cupsctl call).
Please remove the other cupsFilePuts() which add comments from the function and migrate them into sane comments in cupsd.conf.in where it is possible.

> 
>     3. I moved the if else block at the end as well. Linenum checking is used only to not include blank lines between multiline comments.

I would let blank lines between comments as it is - people adds blank lines between comments for readability, and the code will be simpler.

More comments in the review in the moment.

Ankit3002 commented:
@zdohnal  I made the changes as you requested. Please review.
This is how it works:
1. I am checking whether the line i 'm reading is comment or not .
2. if it's a comment than, than i will check for inline comments, and returns the text before '#' as you asked above. ... if it's a simple comment than i simply returns it.
3. if it's not a comment , than I read the configurations according to the earlier function.
4. Also now the function would take care about blank lines as well.

zdohnal commented:
> @zdohnal I made the changes as you requested. Please review. This is how it works:
> 
>     1. I am checking whether the line i 'm reading is comment or not .
> 
>     2. if it's a comment than, than i will check for inline comments, and returns the text before '#' as you asked above. ... if it's a simple comment than i simply returns it.

You can simplify it a lot - inline comments are forbidden, so we can remove them and process as cupsGetConf() did. So in general check for '#' and then iterate over the array towards the start - if there is another text before the #, remove the # and text after # comment, don't return and let the function process the text before #.
In case there is no text, let the function process further - it will remove the leading whitespace and in ```if (buf[0])``` scope you check if the first char is # - if it is - return the buf. 

In the current code there lot of duplicit code, which can be simplified by the algorithm above.

> 
>     3. if it's not a comment , than I read the configurations according to the earlier function.
> 
>     4. Also now the function would take care about blank lines as well.
> 
>     5. I have also placed the function at correct alphabetical order in file.c

There are still more cupsFilePuts() which have to be removed from function and added to cupsd.conf.in if it makes sense.

There are public holidays in my country, I'll be back on Tuesday.

Ankit3002 commented:
> > @zdohnal I made the changes as you requested. Please review. This is how it works:
> > ```
> > 1. I am checking whether the line i 'm reading is comment or not .
> > 
> > 2. if it's a comment than, than i will check for inline comments, and returns the text before '#' as you asked above. ... if it's a simple comment than i simply returns it.
> > ```
> 
> You can simplify it a lot - inline comments are forbidden, so we can remove them and process as cupsGetConf() did. So in general check for '#' and then iterate over the array towards the start - if there is another text before the #, remove the # and text after # comment, don't return and let the function process the text before #. In case there is no text, let the function process further - it will remove the leading whitespace and in `if (buf[0])` scope you check if the first char is # - if it is - return the buf.
> 
> In the current code there lot of duplicit code, which can be simplified by the algorithm above.
> 
> > ```
> > 3. if it's not a comment , than I read the configurations according to the earlier function.
> > 
> > 4. Also now the function would take care about blank lines as well.
> > 
> > 5. I have also placed the function at correct alphabetical order in file.c
> > ```
> 
> There are still more cupsFilePuts() which have to be removed from function and added to cupsd.conf.in if it makes sense.
> 
> There are public holidays in my country, I'll be back on Tuesday.

@zdohnal I simplified the logic as you requested, Please review them when you will be back from holidays, thank you!
The only cupsFilePuts() in adminutil.c are the ones which are used to write missing info.

Ankit3002 commented:
> > > @zdohnal I made the changes as you requested. Please review. This is how it works:
> > > ```
> > > 1. I am checking whether the line i 'm reading is comment or not .
> > > 
> > > 2. if it's a comment than, than i will check for inline comments, and returns the text before '#' as you asked above. ... if it's a simple comment than i simply returns it.
> > > ```
> > 
> > 
> > You can simplify it a lot - inline comments are forbidden, so we can remove them and process as cupsGetConf() did. So in general check for '#' and then iterate over the array towards the start - if there is another text before the #, remove the # and text after # comment, don't return and let the function process the text before #. In case there is no text, let the function process further - it will remove the leading whitespace and in `if (buf[0])` scope you check if the first char is # - if it is - return the buf.
> > In the current code there lot of duplicit code, which can be simplified by the algorithm above.
> > > ```
> > > 3. if it's not a comment , than I read the configurations according to the earlier function.
> > > 
> > > 4. Also now the function would take care about blank lines as well.
> > > 
> > > 5. I have also placed the function at correct alphabetical order in file.c
> > > ```
> > 
> > 
> > There are still more cupsFilePuts() which have to be removed from function and added to cupsd.conf.in if it makes sense.
> > There are public holidays in my country, I'll be back on Tuesday.
> 
> @zdohnal I simplified the logic as you requested, Please review them when you will be back from holidays, thank you! Now,The only cupsFilePuts() in adminutil.c are the ones which are used to write missing info.

@zdohnal Any update on above ?

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +336,131 @@ _cupsFileCheckFilter(
+     if ((ptr = strchr(buf, '#')) != NULL)
+     {
+       int index = (int) (ptr - buf);
+       for(int i=index-1; i>=0; i--)
```
Code style issues (spaces)

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +336,131 @@ _cupsFileCheckFilter(
+       for(int i=index-1; i>=0; i--)
+         if (!_cups_isspace(buf[i]))
+         {
+           buf[index] = '\0';
```
We should put NULL terminator instead of last whitespace in this case.

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +336,131 @@ _cupsFileCheckFilter(
+       * See if there is anything left...
+       */
+
+       if (buf[0] != '#')
```
We have to check whether there is anything in the buffer, and if so, check for # - this big if-else block is unnecessary and add unnecessary indentation - the good practice is to not nest the code too much, the code should be as straight as possible.

```
if (buf[0])
{
  // return comment if any
  if (buf[0] == '#')
    return ptr;

  /*
   * Otherwise grab any value and return...
   */
....
```

zdohnal commented on cups/adminutil.c:
```diff
@@ -800,8 +792,7 @@ cupsAdminSetServerSettings(
-	  cupsFilePuts(temp, "  # Only the owner or an administrator can "
-	                     "cancel a job...\n"
+	  cupsFilePuts(temp, 
 	                     "  <Limit Cancel-Job>\n"
```
Move this into the previous line.

zdohnal commented on cups/adminutil.c:
```diff
@@ -991,8 +951,7 @@ cupsAdminSetServerSettings(
-	cupsFilePuts(temp, "  # Only the owner or an administrator can cancel "
-			   "a job...\n"
+	cupsFilePuts(temp, 
 			   "  <Limit Cancel-Job>\n"
```
Move this into the previous line.

zdohnal commented on cups/file.c:
```diff
@@ -22,6 +22,7 @@
 #include "debug-internal.h"
 #include <sys/stat.h>
 #include <sys/types.h>
+#include <stdbool.h>
```
No need for bool.

zdohnal reviewed (changes requested):
See the comments in the code and there are still cupsFilePuts() which insert comments, which mustn't happen.

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +335,131 @@ _cupsFileCheckFilter(
+     {
+       int index = (int) (ptr - buf);
+       while (index > 0 && _cups_isspace(buf[index]))
+         buf[index--] = '\0';
```
1. setting the NULL terminator in every iteration is redundant - set it only once you know there is a non-whitespace character before #.
2. have you tested the code for a line "       # comment"? What happens for such a line with the current code?
3. what will happen if escaped # appears? - \# - this is not a comment.
4. do you think this can be solved with pointer arithmetic instead of introducing a new variable ```index```?

Help: look at cupsFileGetConf() solves it, think about what is needed in our case, adjust it accordingly and use it here.

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +335,131 @@ _cupsFileCheckFilter(
+  {
+    (*linenum) ++;
+
+    /*
```
We should change and add more comments about what we are doing here and in the if scope.

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +335,131 @@ _cupsFileCheckFilter(
+
+       if (buf[0])
+       {
+         // return comment if any
```
I know that I used it in my pseudocode, but on a second thought it looks out of place when all other one line comments are written as multiline comments. Would you mind turning it into multiline comment and rephrase the comment, so it looks like proper English instead of a personal note :D .

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +335,131 @@ _cupsFileCheckFilter(
 
 
+/*
+ * 'cupsFileGetConfAndComments()' - Get line and comments from a configuration file.
```
The name in the comment doesn't match with the function definition.

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +335,131 @@ _cupsFileCheckFilter(
+
+          return (buf);
+       }
+       else
```
The snippet I've provided in the previous review was pseudocode - not actual code which you can put into the code and have it done, but a code which is supposed to express an idea and has to be adjusted for actual source code.
Additionally I mentioned the if-else structure you introduced in the previous commit is redundant. Think about which part could be removed.

Ankit3002 commented:
@zdohnal Please review the changes that  i made. Thanks!
1. Now the code is using pointer arithmetic, there is no use of index variable.
2. I removed the if-else block and simplified the code.
3. I corrected the function definition

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +335,132 @@ _cupsFileCheckFilter(
+     */
+     if ((ptr = strchr(buf, '#')) != NULL)
+     {
+       ptr--;
```
You don't have to move the pointer, if you check the character before your current position - see cupsFileGetConf

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +335,132 @@ _cupsFileCheckFilter(
+     if ((ptr = strchr(buf, '#')) != NULL)
+     {
+       ptr--;
+       while(ptr >= buf)
```
You will get an address in ```ptr``` which is before ```buf``` pointer with this conditional. Please fix it.

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +335,132 @@ _cupsFileCheckFilter(
+           break;
+         }
+         ptr--;
+       }
```
Missing handling of escaped # .

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +335,132 @@ _cupsFileCheckFilter(
+
+      if (ptr > buf)
+        _cups_strcpy(buf, ptr);
+
```
The redundant part was else block, not whole if-else. Please put the if back.

Ankit3002 replied:
> 

Earlier I have used else block to print blank line as it is. If I just put if block than blank lines won't appears.

Ankit3002 commented:
@zdohnal I resolved the above issues... Now code will also takes care of line like this "--#--something" since this is not an inline comment so it won't get remove.
also i put back the if block , please review Thank you for the clarification above ! 
The code for removing inline comment look like this  
`    /*
     * Remove the inline comment...
     */
     if ((ptr = strchr(buf, '#')) != NULL && _cups_isspace(*(ptr-1)))
     {
        while(ptr > buf)
        {
         // Null-terminate the string after the last non-whitespace
         if(!_cups_isspace(*(ptr-1)))
         {
           *ptr = '\0';
           break;
         }
         ptr--;
        }
     }`

Ankit3002 commented:
> @zdohnal I resolved the above issues... Now code will also takes care of line like this "--#--something" since this is not an inline comment so it won't get remove. also i put back the if block , please review Thank you for the above clarification ! The code for removing inline comment look like this 
` /* * Remove the inline comment... */ if ((ptr = strchr(buf, '#')) != NULL && _cups_isspace(*(ptr-1)))
 { while(ptr > buf) 
{ // Null-terminate the string after the last non-whitespace 
if(!_cups_isspace(*(ptr-1))) { 
*ptr = '\0'; break; } ptr--; 
}
 }`

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +335,136 @@ _cupsFileCheckFilter(
+           return (buf);
+
+        }
+        return (buf);
```
I'm sorry, I didn't realize cupsFileGets() removes the end of lines, so the ```if(buf[0])``` is not met.

Let's ignore the blank lines then - we would have to deal with them by a different cupsFilePuts(), which is not worthy outcome (because we would have to ignore the indent later) - so just remove the return here.

zdohnal replied:
Fixed now - we now support empty lines too.

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +335,136 @@ _cupsFileCheckFilter(
+    /*
+     * Remove the inline comment...
+     */
+     if ((ptr = strchr(buf, '#')) != NULL && _cups_isspace(*(ptr-1)))
```
Please check how cupsFileGetConf() deals with escaped # . This way we would leave '\' in the output file, which is not wanted.

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +335,136 @@ _cupsFileCheckFilter(
+        while(ptr > buf)
+        {
+         // Null-terminate the string after the last non-whitespace
+         if(!_cups_isspace(*(ptr-1)))
```
It would be great if you used ```ptr[-1]``` instead of ```*(ptr-1)``` - it is shorter and in alignment with cupsFileGetConf().

Ankit3002 commented:
@zdohnal I made the changes as you requested . Please review them, The below is how we are taking care of inline comments now.
`  
     if ((ptr = strchr(buf, '#')) != NULL)
     {
        while(ptr > buf)
        {
         // Null-terminate the string after the last non-whitespace, unless the '#' character is escaped by a backslash ('\')
         if(!_cups_isspace(ptr[-1]) && (ptr == buf || ptr[-1] != '\\'))
         {
           *ptr = '\0';
           break;
         }
         ptr--;
        }
     }
`

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +335,135 @@ _cupsFileCheckFilter(
+        */
+        if(buf[0])
+        {
+          if (buf[0] == '#')
```
I've thought a bit about dealing with the empty lines in the yesterday evening. We can return ```buf``` here together with ```buf[0] == '#'```, so the condition would look like:

```
/*
 * If we got empty line or comment, return it...
 */
if (!buf[0] || buf[0] == '#')
  return buf;
```

and we can remove the ```if(buf[0])``` before that.

We just need to add a new ```if else if``` in cupsAdminSetServerSettings():

``` 
while (_cupsFileGetConfandComments(cupsd, line, sizeof(line), &value, &linenum))
{
 /*
  *  Preserve empty lines
  */
  if (!line[0])
    cupsFilePuts(temp, "\n");
  else if ((!_cups_strcasecmp(line, "Port") || !_cups_strcasecmp(line, "Listen")) &&
```

This should do the trick - WDYT? I guess you might show something similar in the past, but I didn't get it in the past, sorry for that :D .

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +335,135 @@ _cupsFileCheckFilter(
+  {
+    (*linenum) ++;
+
+    /*
+     * Remove the inline comment...
+     */
+     if ((ptr = strchr(buf, '#')) != NULL)
+     {
+        while(ptr > buf)
+        {
+         // Null-terminate the string after the last non-whitespace, unless the '#' character is escaped by a backslash ('\')
+         if(!_cups_isspace(ptr[-1]) && (ptr == buf || ptr[-1] != '\\'))
+         {
+           *ptr = '\0';
+           break;
+         }
+         ptr--;
+        }
+     }
```
Incorrect indentation - every line (except for line with opening multiblock comment ```/*```) is indented by even number of whitespaces - (2 spaces -> 4 spaces -> 6 spaces -> 1 tab -> 1 tab and 2 spaces etc). The initial indentation here is 5, which is odd number.

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +335,135 @@ _cupsFileCheckFilter(
+          {
+           /*
+            * Have a value, skip any other spaces...
+            */
+
+            while (_cups_isspace(*ptr))
+              *ptr++ = '\0';
+
+            if (*ptr)
+              *value = ptr;
+
+           /*
+            * Strip trailing whitespace and > for lines that begin with <...
+            */
+
+            ptr += strlen(ptr) - 1;
+
+            if (buf[0] == '<' && *ptr == '>')
+              *ptr-- = '\0';
+            else if (buf[0] == '<' && *ptr != '>')
+            {
+             /*
+              * Syntax error...
+              */
+
+              *value = NULL;
+              return (buf);
+            }
+
+            while (ptr > *value && _cups_isspace(*ptr))
+              *ptr-- = '\0';
+          }
+
+          /*
+           * Return the line...
+           */
+
+           return (buf);
+
+        }
```
The indentation is incorrect till this part - incorrect number of whitespaces and no tabs were used.

zdohnal reviewed (changes requested):
I've reviewed the code from code style POV - tabs weren't used on places where they should be and the indentation is shifted since the second loop.

Would you mind looking into it?

Ankit3002 commented:
@zdohnal I made the changes as you requested. Please review them.

Ankit3002 commented:
> @zdohnal I made the changes as you requested. Please review them.

@zdohnal  Any update on this ?

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +335,165 @@ _cupsFileCheckFilter(
+         * Remove the backslash and continue searching for unescaped '#'...
+         */
+        _cups_strcpy(ptr-1, ptr);
+        ptr = strchr(ptr+1, '#');
```
Good thinking here! Could you rework the functionality (recursively checking for #, because it can indicate forbidden inline comment or escaped #) into a static local function, which can be called recursively? So we can just do once:

```
if ((ptr = strchr(buf, '#')) != NULL)
  cups_handle_hash(ptr, buf);
```

this way you don't have to check for escaped # in the loop again, and the while loop can be in ```else``` scope, so we spare one conditional check.

There are code style issues as well (ptr-1 -> ptr - 1 , the comments are shifted by one space - look into other code in the file to see how they should look).

zdohnal commented on cups/file.c:
```diff
@@ -335,6 +335,165 @@ _cupsFileCheckFilter(
+    if (ptr > buf)
+      _cups_strcpy(buf, ptr);
+
+    /*
```
Please change the comment here - now we don't return only comment, but in case of empty line as well.

Ankit3002 commented:
> @zdohnal I made the changes as you requested. Please review them.

@zdohnal Any Update on this ?

zdohnal reviewed (changes requested):
Hi,

good thinking about possibility of having more # on one line! IMO it would be great if that could be handled by a function, so we can call it recursively there. Would you mind looking into it?
There are other code style issues, please look into them as well.

Thank you in advance!

I'm going to vacation for next two weeks and I'll return on 15th May, so I won't be available for reviews till I return, so see you in two weeks!

zdohnal commented:
Hi @Ankit3002 ,

have you got time to look into the PR?

Ankit3002 commented:
> Hi @Ankit3002 ,
> 
> have you got time to look into the PR?

Yes, I will push the changes today.

Ankit3002 commented:
@zdohnal Please review the changes that I made ...

Ankit3002 commented:
> @zdohnal Please review the changes that I made ...

@zdohnal Any Update on this ?

michaelrsweet commented:
Definitely not for 2.4.x, *maybe* for 2.5.

@zdohnal Will leave this review process to you since you've been looking at this code more than me lately...

michaelrsweet commented:
@zdohnal Is this still something you are looking at? I am very leery of making changes to this code...
